File: Data Structures & Algorithms/combination-target-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        let ans = [];
        let curr = [];
        this.backtrack(nums, target, ans, curr, 0);
        return ans;
    }

    backtrack(nums, target, ans, curr, index){
        if(target === 0) ans.push([...curr])
        else if(target < 0 || index >= nums.length) return;
        else {
            curr.push(nums[index])
            this.backtrack(nums, target - nums[index], ans, curr, index)
            curr.pop()
            this.backtrack(nums, target, ans, curr, index + 1)
        }
    }
}

File: Data Structures & Algorithms/subsets/submission-0.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(0 ,nums, new ArrayList<>(), result);
        return result;
    }

    private void backtrack(int index, int[] nums, List<Integer> subset, List<List<Integer>> result) {
            result.add(new ArrayList(subset));

        for (int i = index; i < nums.length; i++) {
            subset.add(nums[i]);
            backtrack(i + 1, nums, subset,  result);
            subset.remove(subset.size() - 1);
        }
    }
}
